train.lua

```diff
@@ -87,98 +79,144 @@ do
       optnet.optimizeMemory(net, sample_input, {inplace = false, mode = 'training'})
    end
 
-   model:add(utils.makeDataParallelTable(net, opt.nGPU))
 end
+model:add(utils.makeDataParallelTable(net, opt.nGPU))
+cast(model)
 
-local function log(t) print('json_stats: '..json.encode(tablex.merge(t,opt,true))) end
-
-print('Will save at '..opt.save)
-paths.mkdir(opt.save)
-
-local parameters,gradParameters = model:getParameters()
-
-opt.n_parameters = parameters:numel()
-print('Network has ', parameters:numel(), 'parameters')
-
-print(c.blue'==>' ..' setting criterion')
-local criterion = nn.CrossEntropyCriterion():cuda()
-
--- a-la autograd
-local f = function(inputs, targets)
-   model:forward(inputs)
-   local loss = criterion:forward(model.output, targets)
-   local df_do = criterion:backward(model.output, targets)
-   model:backward(inputs, df_do)
-   return loss
+local function hflip(x)
+   return torch.random(0,1) == 1 and x or image.hflip(x)
 end
 
-print(c.blue'==>' ..' configuring optimizer')
-local optimState = tablex.deepcopy(opt)
-
-
-function train()
-  model:training()
-
-  local targets = torch.CudaTensor(opt.batchSize)
-  local indices = torch.randperm(provider.trainData.data:size(1)):long():split(opt.batchSize)
-  -- remove last element so that all minibatches have equal size
-  indices[#indices] = nil
-
-  local loss = 0
+local function randomcrop(x)
```

Comment:
I think the overhead is small for now, will wait for torchnet-vision to fix this